### ISSUE TITLE
 Possible performance improvements for sqflite plugin

### ISSUE THREAD

**Issue by mehmetf** [tekartik/sqflite#459]
While investigating cold-start slowness of SQLite databases, we noticed that a fair amount of time was spent in the database open step. Digging deeper, in https://github.com/tekartik/sqflite/blob/5cd8b1bc8e11dfe8b7485e20fb5c345e8431162f/sqflite_common/lib/src/database_mixin.dart#L691-L718, a transaction is created to perform the `onCreate/onUpgrade/onDowngrade` steps. Version is retrieved during that transaction. On a pixel 3, this transaction appears to take ~50ms.

In most cases, the database version is the same and these 3 steps don't need to be performed, so only the "PRAGMA user_version" is run within the transaction. Could some time be saved if the version is checked outside the transaction and then not perform the transaction if there is no need to perform any of the onCreate/onUpgrade/onDowngrade steps?

**Comment by alextekartik** [comment]
I see indeed. Like you say most of the time no migration (onCreate/onUpgrade/onDowngrade) is  needed.

I will look into that. The likely change:
- Version is read before the transaction
- if no version change, no transaction is created
- if the version has changed, the transaction is created (and to handle edge cases, the version will be read again during the transaction)

**Comment by alextekartik** [comment]
I published sqflite_common 1.0.2 (I have not updated sqflite itself so a force upgrade is needed) with the fix. Thanks for spotting this and for the proposed change. Let me know if that works for you and if you see some improvments.

As a side note, on Android, I use a low priority thread for handling SQL request. I have not done extensive performance testing to compare with a normal thread priority so if someone is doing performance testing with exact timing, I would love to know if the number changes.

One way to force NORMAL_PRIORITY (to call first before any other calls):
```dart
await Sqflite.devSetOptions(
            SqfliteOptions()..androidThreadPriority = 0);
```

(yes it is full of deprecated warning - on purpose).

**Comment by alextekartik** [comment]
(closing assuming this is fixed for you)
